[LBSE] Add new tests that cover repainting after transform changes

nikolaszimmermann opened the pull request:
#### c417abc93c3b16d7e504bf27e1e3b4839820a6e6
<pre>
[LBSE] Add new tests that cover repainting after transform changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=249142">https://bugs.webkit.org/show_bug.cgi?id=249142</a>

Reviewed by Rob Buis.

Add tests that cover repainting in different scenarios: foreignObject, image, text, shapes
either directly transformed themselves, or via parents. That was used in bootstrapping
the LBSE optimization to avoid relayouts, besides the existing tests.

Unfortunately, the existing repaint tests are not sufficient to test repainting visually,
because the existing svg/repaint tests rely on pixel testing, which is off by default.

Therefore, provide modern repaint test in two forms:
a) repaint rects as text (platform agnostic, except for SVG &lt;text&gt; - platform specific results are expected)
b) repaint reftests (to rule out repainting artefacts, etc.)

No change in functionality, only new tests.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/foreign-object-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/foreign-object-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/image-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/image-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/shape-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/shape-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/text-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/win/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-container-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-transform.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-container-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-transform.html: Added.
* LayoutTests/svg/repaint/inner-svg-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/inner-svg-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/inner-svg-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/inner-svg-repainting-after-modifying-transform.html: Added.
* LayoutTests/svg/repaint/resources/reftest-repaint.js: Added.
(runRepaintRefTest):
(callRepaintRefTest):
(finishRefTest):
* LayoutTests/svg/repaint/shape-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-container-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-transform.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-transform.html: Added.

Canonical link: <a href="https://commits.webkit.org/259243@main">https://commits.webkit.org/259243@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29c02bbfa73d270e04c83388d252510c8c88e04c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104242 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13331 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37154 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113453 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173743 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14408 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4244 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96466 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112505 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110011 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11099 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94158 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38756 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92959 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25769 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80421 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6692 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27129 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6826 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3683 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12841 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46682 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6374 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8612 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis reviewed (approved):
LGTM. Some tests can maybe become WPT tests, but that can be done later.

nikolaszimmermann committed:
[LBSE] Add new tests that cover repainting after transform changes
https://bugs.webkit.org/show_bug.cgi?id=249142

Reviewed by Rob Buis.

Add tests that cover repainting in different scenarios: foreignObject, image, text, shapes
either directly transformed themselves, or via parents. That was used in bootstrapping
the LBSE optimization to avoid relayouts, besides the existing tests.

Unfortunately, the existing repaint tests are not sufficient to test repainting visually,
because the existing svg/repaint tests rely on pixel testing, which is off by default.

Therefore, provide modern repaint test in two forms:
a) repaint rects as text (platform agnostic, except for SVG <text> - platform specific results are expected)
b) repaint reftests (to rule out repainting artefacts, etc.)

No change in functionality, only new tests.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/ios/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/foreign-object-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/foreign-object-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/image-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/image-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/shape-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/shape-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/text-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/platform/mac/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/win/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-container-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/foreign-object-repainting-after-modifying-transform.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-container-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/image-repainting-after-modifying-transform.html: Added.
* LayoutTests/svg/repaint/inner-svg-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/inner-svg-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/inner-svg-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/inner-svg-repainting-after-modifying-transform.html: Added.
* LayoutTests/svg/repaint/resources/reftest-repaint.js: Added.
(runRepaintRefTest):
(callRepaintRefTest):
(finishRefTest):
* LayoutTests/svg/repaint/shape-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-container-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/shape-repainting-after-modifying-transform.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-container-transform-expected.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-container-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-container-transform.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-transform-expected.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-transform-repaint-rects-expected.txt: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-transform-repaint-rects.html: Added.
* LayoutTests/svg/repaint/text-repainting-after-modifying-transform.html: Added.

Canonical link: https://commits.webkit.org/259243@main
